pymmcore_plus/mda_writers/_writers.py

```diff
@@ -0,0 +1,65 @@
+from pathlib import Path
+from typing import Tuple
+
+import numpy as np
+from typing_extensions import Protocol  # typing extensions for 3.7 support
+from useq import MDAEvent, MDASequence
+
+try:
+    import tifffile
+except ImportError:
+    tifffile = None
+
+
+class MDAWriter(Protocol):
+    def sequenceStarted(self, seq: MDASequence) -> None:
+        ...
+
+    def sequenceCanceled(self, seq: MDASequence) -> None:
+        ...
+
+    def sequenceFinished(self, seq: MDASequence) -> None:
+        ...
+
+    def frameReady(self, img: np.ndarray, event: MDAEvent) -> None:
+        ...
+
+    def sequencePauseToggled(self, paused: bool) -> None:
+        ...
+
+    @staticmethod
+    def get_axis_order(seq: MDASequence) -> Tuple[str]:
+        """Get the axis order using only axes that are present in events."""
+        event = next(seq.iter_events())
+        event_axes = list(event.index.keys())
+        return tuple(a for a in seq.axis_order if a in event_axes)
```

Comment:
This is something I found necessary as you can run into errors when querying an even for it's index if the axis order contains a dimension that isn't actually used to create an event. e.g.:

```python
mda = MDASequence(
    # stage_positions=[(100, 100, 30), (200, 150, 35)],
    channels=["DAPI", "FITC"],
    time_plan={'interval': 1, 'loops': 20},
    z_plan={"range": 4, "step": 0.5},
    axis_order='tpcz',
)
```

Reply:
yeah, I've run into this a lot as well.  it's annoying.  We should think about how to fix this upstream as well.